include/dali/core/cuda_utils.h

```diff
@@ -93,6 +93,16 @@ int MaxThreadsPerBlock(KernelFunction *f) {
   return max_block_size[device];
 }
 
+inline int GetSmCount() {
+  static int count = 0;
+  if (!count) {
+    cudaDeviceProp prop;
+    CUDA_CALL(cudaGetDeviceProperties(&prop, 0));
+    count = prop.multiProcessorCount;
+  }
+  return count;
+}
```

Comment:
```suggestion
inline int GetSmCount(int device_id = -1) {
Sorry, but that's not sufficient. There can be more than one device - and indeed, more than one type of device.
```suggestion
  if (device_id < 0)
    CUDA_CALL(cudaGetDevice(&device_id));
  static int dev_count = []() {
    int ndevs = 0;
    CUDA_CALL(cudaGetDeviceCount(&ndevs));
  }();
  static unique_ptr<int[]> count(new int[dev_count]());  // this should be zero-initialized
  if (!count[device_id]) {
    cudaDeviceProp prop;
    CUDA_CALL(cudaGetDeviceProperties(&prop, 0));
    count[device_id] = prop.multiProcessorCount;
  }
  return count[device_id];
```
}
```

Reply:
I've fixed this as you suggested in the latest commit with a small difference. I decided to use `vector` instead of `unique_ptr` to an array, because I think it's more obvious what it does then. If there's some advantage of the `unique_ptr` approach, of course I'll change it.